packages/uniforms/src/BaseForm.js

```diff
@@ -238,10 +247,30 @@ export default class BaseForm extends Component {
             event.stopPropagation();
         }
 
-        return Promise.resolve(
-            this.props.onSubmit &&
-            this.props.onSubmit(this.getModel('submit'))
-        ).then(
+        let submitting;
+        try {
+            const res = this.props.onSubmit && this.props.onSubmit(this.getModel('submit'));
+            submitting = Promise.resolve(res);
+
+            // Do not change the `submitting` state if onSubmit is not async
+            if (res && isFunction(res.then)) {
+                this.setState({submitting: true});
+                submitting = submitting.then(
+                    val => {
+                        this.setState({submitting: false});
+                        return val;
+                    },
+                    err => {
+                        this.setState({submitting: false});
+                        throw err;
+                    }
+                );
+            }
+        } catch (error) {
+            submitting = Promise.reject(error);
+        }
+
+        return submitting.then(
```

Comment:
Three things here:
* What error exactly is possible here? `onSubmit` should return a `Promise`, so I don't see an error possibility here.
* I'd rather simplify it with calling `setState` before the first resolve and chain one more `.then`, which will reset the state. I'm not sure about the polyfill, but I think that `Promise.prototype.finally` is the way to go.
* Please use full names: I won't accept `res` nor `err`.

Reply:
- The try/catch was introduced (in 53a9690) because `onSubmit` is a user provided function so we don't really know how it will behave. Because uniforms tracks the success/failure of `onSubmit`, it felt natural to also account for this failure mode and ensure the validity of uniform's internal state. That being said, the precedent has been set elsewhere that the user is responsible for the behaviour of their callbacks, so I'm happy to remove the try/catch.

- Conditionally setting the state like this (added in 1b4e727) is an optimisation to avoid an unnecessary re-render when `onSubmit` is not async. From [the docs](https://reactjs.org/docs/react-component.html#setstate) "`setState()` will always lead to a re-render unless `shouldComponentUpdate()` returns `false`", so setting `submitting` to `true` then to `false` will cause a re-render even if both calls are synchronous. I think a promise `then` callback will always happen asynchronously, even if the promise has already resolved, so it could even cause two unnecessary re-renders. It's up to whether the added complexity is worth it.

- Sure.

Reply:
* Let's remove it then - we've never accepted such behaviour and I can't see an use case here. Most of `onSubmit`s will be asynchronous anyway.
* You can always return `null` to prevent the render. Anyway, it's OK.

Reply:
- That is sensible. I have removed the try/catch.